File: ts/33.search-in-rotated-sorted-array.ts
/*
 * @lc app=leetcode id=33 lang=typescript
 *
 * [33] Search in Rotated Sorted Array
 */

// @lc code=start
function search(nums: number[], target: number): number {
  const findRotateIndex = (left: number, right: number) => {
    if (nums[left] < nums[right]) return 0;

    while (left <= right) {
      let pivot = Math.floor((left + right) / 2);
      if (nums[pivot] > nums[pivot + 1]) {
        return pivot + 1;
      } else {
        if (nums[pivot] < nums[left]) {
          right = pivot - 1;
        } else {
          left = pivot + 1;
        }
      }
    }
    return 0;
  };

  const binarySearch = (left: number, right: number) => {
    while (left <= right) {
      let pivot = Math.floor((left + right) / 2);
      if (nums[pivot] === target) {
        return pivot;
      } else {
        if (target < nums[pivot]) {
          right = pivot - 1;
        } else {
          left = pivot + 1;
        }
      }
    }

    return -1;
  };

  let n = nums.length;

  if (n === 1) return nums[0] === target ? 0 : -1;

  let rotateIndex = findRotateIndex(0, n - 1);

  if (nums[rotateIndex] === target) return rotateIndex;

  if (rotateIndex === 0) return binarySearch(0, n - 1);

  if (target < nums[0]) return binarySearch(rotateIndex, n - 1);

  return binarySearch(0, rotateIndex);
}
// @lc code=end

File: go/33.search-in-rotated-sorted-array/33.search-in-rotated-sorted-array.go
/*
 * @lc app=leetcode id=33 lang=golang
 *
 * [33] Search in Rotated Sorted Array
 */

package leetcode

// @lc code=start
func search(nums []int, target int) int {
	n := len(nums) - 1
	left, right := 0, n

	for left <= right {
		mid := (left + right) / 2

		if nums[mid] == target {
			return mid
		}

		// left part is ordered
		if nums[mid] >= nums[left] {
			// target exists between left and mid
			if target >= nums[left] && target < nums[mid] {
				right = mid - 1
			} else {
				left = mid + 1
			}
		} else {
			// target exists between mid and right
			if target <= nums[right] && target > nums[mid] {
				left = mid + 1
			} else {
				right = mid - 1
			}
		}
	}

	return -1
}

// @lc code=end
